pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java

```diff
@@ -279,15 +279,21 @@ public void redeliverUnacknowledgedMessages(Consumer consumer, long consumerEpoc
     }
 
     private synchronized void internalRedeliverUnacknowledgedMessages(Consumer consumer, long consumerEpoch) {
+        // broker side epoch is smaller than consumer epoch, so don't need to handle this redeliver request
+        if (consumerEpoch < consumer.getConsumerEpoch()) {
+            log.warn("[{}-{}] Ignoring redeliverUnacknowledgedMessages since broker epoch [{}] is smaller than "
+                            + "consumer epoch [{}]",
+                    name, consumer, consumer.getConsumerEpoch(), consumerEpoch);
+            return;
+        }
 
-        if (consumerEpoch > consumer.getConsumerEpoch()) {
-            if (log.isDebugEnabled()) {
-                log.debug("[{}-{}] Update epoch, old epoch [{}], new epoch [{}]",
-                        name, consumer, consumer.getConsumerEpoch(), consumerEpoch);
-            }
-            consumer.setConsumerEpoch(consumerEpoch);
+        if (log.isDebugEnabled()) {
+            log.debug("[{}-{}] Update epoch, old epoch [{}] new epoch [{}]",
+                    name, consumer, consumer.getConsumerEpoch(), consumerEpoch);
         }
 
+        consumer.setConsumerEpoch(consumerEpoch);
```

Comment:
This doesn't make sense to me. Why is the broker side epoch updated here?

Reply:
this means that client consumer invoke redeliverUnacknowledgedMessages, client consumer hope that broker can next send messages with consumerEpoch.

Reply:
I get it now. It seems that the code was just refactored in this PR and the logic wasn't really changed in this method.